membersonlyevent.php

```diff
@@ -585,56 +585,15 @@ function membersonlyevent_civicrm_entityTypes(&$entityTypes) {
 }
 
 /**
- * Implements hook_civicrm_pre().
+ * Implements hook_civicrm_copy().
  *
- * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_pre/
+ * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_copy/
  */
-function membersonlyevent_civicrm_pre($op, $objectName, $id, &$params) {
+function membersonlyevent_civicrm_copy($objectName, &$object) {
   $listeners = [
-    new CRM_MembersOnlyEvent_Hook_Pre_Event(),
+    new CRM_MembersOnlyEvent_Hook_Copy_Event(),
```

Comment:
maybe you can do this instead : 

```

function membersonlyevent_civicrm_copy($objectName, &$object) {
 if ($objectName == 'Event') { 
  $templateId = CRM_Utils_Request::retrieve('template_id', 'Int');

  $listener = new CRM_MembersOnlyEvent_Hook_Copy_Event($object->id, $templateId),
  $listener->handle();
  }
}
```


the benefits is that you won't need to initialize the class every time this hook is called whether it  is from the right context or not, and that you can test the class like that.

Reply:
> and we may want to do something else without templated for copy event hook.

if you want to account for that I can see two options : 

1 - make `createMemberOnlyEventFromEventTemplate` method public and call it instead of calling `handle` and thus you can pass the template id to the method direclty.

2- separate `createMemberOnlyEventFromEventTemplate` method into a new class and pass the template id to it and then you can test this new class , where the main hook class CRM_MembersOnlyEvent_Hook_Copy_Event will  stay (and still be called from the extension.php file) but will serve to group the actions that will be done when the copy hook is called